### PR TITLE
Generalized material spawning

### PR DESCRIPTION
**Pull request by Elijahrane**
## About the PR 
It all seems to work fine, I'm slightly unhappy with how some of it looks and I'll probably do a follow up cleanup PR removing StackPrototype.

**Screenshots**

https://user-images.githubusercontent.com/60792108/200721027-bfb4a155-c6e8-437d-aa17-9b0274fc0ced.mp4

**Changelog**

:cl: Rane
- add: All machines that can hold materials no longer destroy them on deconstruction.

### CODE REVIEW

**Comment by EmoGarbage404** [comment]
if you're attaching entities to materials you really need to just bite the bullet and get rid of tags for ingots/sheets/ore and use separate materials for each.

**Comment by Elijahrane** [comment]
> if you're attaching entities to materials you really need to just bite the bullet and get rid of tags for ingots/sheets/ore and use separate materials for each.

I don't see why, really. Right now I'm deciding if removing stackprototype is within the scope of this PR

File: Content.Shared/Materials/MaterialPrototype.cs [hunk 1]
```diff
@@ -27,8 +24,16 @@ public sealed class MaterialPrototype : IPrototype, IInheritingPrototype
...
         public string ID { get; } = default!;
 
-        [DataField("stack", customTypeSerializer:typeof(PrototypeIdSerializer<StackPrototype>))]
-        public string? StackId { get; } = null;
+        /// <summary>
+        ///     For material storage to be able to convert back and forth
+        ///     between the material and physical entities you can carry,
+        ///     include which stack we should spawn by default.
+        /// </summary>
+        [ViewVariables]
```
Comment (ElectroJr) [inline Content.Shared/Materials/MaterialPrototype.cs:32]: [ViewVariables] is implicit for any [DataField]

File: Content.Shared/Materials/MaterialPrototype.cs [hunk 2]
```diff
@@ -27,8 +24,16 @@ public sealed class MaterialPrototype : IPrototype, IInheritingPrototype
...
+        [DataField("stackEntity", customTypeSerializer:typeof(PrototypeIdSerializer<EntityPrototype>))]
+        public string StackEntity { get; } = "";
+
+        private string _name = string.Empty;
```
Comment (ElectroJr) [inline Content.Shared/Materials/MaterialPrototype.cs:36]: this is unused. And why does the name data-field have a private set?

File: Content.Server/Stack/StackSystem.cs [hunk 3]
```diff
@@ -87,24 +88,89 @@ public EntityUid Spawn(int amount, StackPrototype prototype, EntityCoordinates s
...
+            if (amount <= 0)
+                return list;
+
+            // At least 1 is being spawned, we'll use the first to extract otherwise inaccessible information
+            // ??TODO??: Indexing the entity proto and extracting from its component registry could possibly be better?
+            // it doesn't look like it would save LOC even compressing this to a single loop and I'm not sure what other issues it might introduce
+            var firstSpawn = Spawn(materialProto.StackEntity, coordinates);
+
+            if (!TryComp<StackComponent>(firstSpawn, out var stack) || stack.StackTypeId == null)
+                return list;
```
Comment (ElectroJr) [inline Content.Server/Stack/StackSystem.cs:106]: Is there a test that asserts that every material prototype entity has a stack component?

If there isn't this should definitely log errors & properly handle deleting the spawned entity.

File: Content.Server/Stack/StackSystem.cs [hunk 4]
```diff
@@ -87,24 +88,89 @@ public EntityUid Spawn(int amount, StackPrototype prototype, EntityCoordinates s
...
+
+            // yes every time you want to get a max count on a stack component you have to index a prototype
+            if (!_prototypeManager.TryIndex<StackPrototype>(stack.StackTypeId, out var stackProto) || stackProto.MaxCount == null)
+                return list;
```
Comment (ElectroJr) [inline Content.Server/Stack/StackSystem.cs:110]: This should use `SharedStackSystem.GetMaxCount()`

File: Content.Server/Stack/StackSystem.cs [hunk 5]
```diff
@@ -87,24 +88,89 @@ public EntityUid Spawn(int amount, StackPrototype prototype, EntityCoordinates s
...
+                    var entity = Spawn(materialProto.StackEntity, coordinates);
+                    list.Add(entity);
+                    var nextStack = Comp<StackComponent>(entity);
+
+                    SetCount(entity, (amount / materialPerStack), nextStack);
+                    amount = 0;
+                }
             }
-            return spawnedEnts;
+            return list;
```
Comment (ElectroJr) [inline Content.Server/Stack/StackSystem.cs:158]: firstSpawn never gets added to the list

File: Content.Server/Stack/StackSystem.cs [hunk 6]
```diff
@@ -87,24 +88,89 @@ public EntityUid Spawn(int amount, StackPrototype prototype, EntityCoordinates s
...
-                var entity = Spawn(entityPrototype, spawnPosition);
-                spawnedEnts.Add(entity);
-                var countAmount = Math.Min(maxCountPerStack, amount);
-                SetCount(entity, countAmount);
-                amount -= countAmount;
+                if (amount > materialPerMaxCount)
+                {
+                    var entity = Spawn(materialProto.StackEntity, coordinates);
+                    list.Add(entity);
+                    var nextStack = Comp<StackComponent>(entity);
```
Comment (ElectroJr) [inline Content.Server/Stack/StackSystem.cs:143]: This block can just be moved out of the if statement to avoid duplication

**Review by ElectroJr — CHANGES_REQUESTED** [review]
.

File: Content.Server/Stack/StackSystem.cs (hunk 3 above)
Comment (Elijahrane) [reply Content.Server/Stack/StackSystem.cs:106]: I wrote a test

**Review by AJCM-git — APPROVED** [review]

File: Content.IntegrationTests/Tests/Materials/MaterialTests.cs [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+
+                foreach (var proto in allMaterialProtos)
+                {
+                    if (proto.StackEntity == "")
+                        continue;
+
+                    var spawned = entityManager.SpawnEntity(proto.StackEntity, coords);
+
+                    Assert.That(entityManager.HasComponent<StackComponent>(spawned),
+                        $"{proto.ID} 'stack entity' {proto.StackEntity} has the stack component");
```
Comment (ElectroJr) [inline Content.IntegrationTests/Tests/Materials/MaterialTests.cs:49]: These messages are logged when the assert fails, so they state the exact opposite. But ill just fix this when fixing conflicts in 13326

**Review by ElectroJr — APPROVED** [review]

**Comment by EmoGarbage404** [comment]
Does this still have the issue where it doesnt work properly for materials that are used in different contexts? For example, ore placed into a processor in ore form shouldnt be returned as an ingot on deconstruction.

**Comment by ElectroJr** [comment]
AFAIK this is already just a general problem for anything using material storage unless they somehow decide to track the input entity prototypes? So either material storage needs to be reworked or the ore processor needs to just store the ore instead of directly turning it into materials. Though I guess in the meantime the drop-on-deconstruct should be disabled for some machines?

**Pull request merged**